### ISSUE TITLE
Implement advanced_signal<> class

### ISSUE THREAD

**Issue by sergey-shambir** [ispringteam/FastSignals#5]
This class should implement advanced features with their trade-offs.
It should:
* return `advanced_connection` from `connect` method
* support `shared_connection_block` in `advanced_connection` class
* support `extended_connect` to allow slot access it's connection
Class `advanced_scoped_connection` should be implemented to as RAII for the `advanced_connection`.

**Comment by Warboss-rus** [comment]
Implemented feature in "`advanced_connection`" branch. I opted to create new tagged method in signal class instead of creating entirely new class for blockable slots. This allows some slots to be blockable, while others use existing code without perfomance penalty for the ability to block. But there is a limitation: only signals that return void can connect using new method. The other limitation is that `advanced_connection` cannot be stored in regular `scoped_connection` (constructor explicitly deleted) and requires use of `advanced_scoped_connection`. To block the slot, use `shared_connection_block` class. Example:
```signal<void(int)> event;
auto conn = event.connect([&](int gotValue) {
	CHECK(gotValue == 123);
}, advanced_tag{});
shared_connection_block block(conn);
event(123); //callback will not be called
```
Code review is required
